pkg/dns/treecache.go

```diff
@@ -18,108 +18,68 @@ package dns
 
 import (
 	"bytes"
-	"crypto/md5"
 	"encoding/json"
-	"fmt"
-	"io/ioutil"
-	"os"
-	"path"
-	"reflect"
 	"strings"
 	"sync"
 )
 
-const (
-	dataFile = "data.dat"
-	crcFile  = "data.crc"
-)
-
 type object interface{}
 
 type TreeCache struct {
 	ChildNodes map[string]*TreeCache
 	Entries    map[string]interface{}
-	m          *sync.RWMutex
+	rwLock     *sync.RWMutex
 }
 
-func NewTreeCache() *TreeCache {
+func NewTreeCache(m *sync.RWMutex) *TreeCache {
 	return &TreeCache{
 		ChildNodes: make(map[string]*TreeCache),
 		Entries:    make(map[string]interface{}),
-		m:          &sync.RWMutex{},
+		rwLock:     m,
 	}
 }
 
-func Deserialize(dir string) (*TreeCache, error) {
-	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
-	if err != nil {
-		return nil, err
-	}
-	var hash []byte
-	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
-	if err != nil {
-		return nil, err
-	}
-	if !reflect.DeepEqual(hash, getMD5(b)) {
-		return nil, fmt.Errorf("Checksum failed")
+func (cache *TreeCache) Serialize() (string, error) {
+	if cache.rwLock != nil {
```

Comment:
blech.  This encourages me to be confused about locking.  As far as I can tell, there is no place that calls this on anything but the root.  Just skip the lock here and add a comment that this assumes it is operating on a locked tree.